pkgs/stdenv/generic/setup.sh

```diff
@@ -71,21 +73,18 @@ _callImplicitHook() {
     set -u
     local def="$1"
     local hookName="$2"
-    case "$(type -t "$hookName")" in
-        (function|alias|builtin)
-            set +u
-            "$hookName";;
-        (file)
-            set +u
-            source "$hookName";;
-        (keyword) :;;
-        (*) if [ -z "${!hookName:-}" ]; then
-                return "$def";
-            else
-                set +u
-                eval "${!hookName}"
-            fi;;
-    esac
+    if declare -F "$hookName" > /dev/null; then
+        set +u
+        "$hookName"
+    elif type -p "$hookName" > /dev/null; then
+        set +u
+        source "$hookName"
+    elif [ -n "${!hookName:-}" ]; then
+        set +u
+        eval "${!hookName}"
+    else
+        return "$def"
+    fi
```

Comment:
Comment on `_callImplicitHook`:
This is tricky. Following assumption is taken:
`$hookName` can't be an alias, builtin or keyword. It's either a function, or a file, or a variable, or the default value is returned.

<details><summary>Test example</summary><p>

```sh
#!/usr/bin/env bash

set -eu
set -o pipefail

orig() {
    local hookName=$1
    case "$(type -t "$hookName")" in
        (function|alias|builtin) echo function;;
        (file) echo file;;
        (keyword) :;;
        (*) if [ -z "${!hookName:-}" ]; then
            echo default
        else
            echo variable
        fi;;
    esac
}

new() {
    local hookName=$1
    if declare -F "$hookName" > /dev/null; then
        echo function
    elif type -p "$hookName" > /dev/null; then
        echo file
    elif [ -n "${!hookName:-}" ]; then
        echo variable
    else
        echo default
    fi
}

dummyfunc() { :; }
dummyvar=value

echo orig
orig dummyfunc
orig ./05.sh
orig bash
orig dummyvar
orig nonexistent
echo
echo new
new dummyfunc
new ./05.sh
new bash
new dummyvar
new nonexistent
```
</p></summary>